server_list/models.py

```diff
@@ -0,0 +1,266 @@
+		else:
+			server = Server()
+			server.update(obj, True)
+			db.session.add(server)
+		return server
+
+	def update(self, obj, initial=False):
+		now = datetime.now()
+		action = obj.get("action", "start")
+
+		assert action != "delete"
+
+		if "clients_list" in obj:
+			num_clients = len(obj["clients_list"])
+		else:
+			num_clients = obj["clients"]
+
+		if initial:
+			# Values set only when the server is first created
+			assert action == "start"
+			self.world_uuid = obj.get("world_uuid")
+			self.clients_top = num_clients
+			self.total_uptime = 0
+		else:
+			self.clients_top = max(self.clients_top, num_clients)
+
+		if action == "start":
+			# Fields updated only on startup
+			self.start_time = now
+			self.mods = "\n".join(obj.get("mods", []))
+			self.mapgen = obj.get("mapgen")
+			self.default_privs = obj.get("privs")
+			self.is_dedicated = obj.get("dedicated", False)
+			self.rollback_enabled = obj.get("rollback", False)
+			self.can_see_far_names = obj.get("can_see_far_names", False)
+
+		self.online = True
+
+		self.address = obj["address"]
+		self.port = obj.get("port", 30000)
+
+		self.announce_ip = obj["ip"]
+
+		self.server_id = obj.get("server_id")
+
+		self.clients = "\n".join(obj["clients_list"])
+		self.clients_max = obj["clients_max"]
+
+		self.game_time = obj["game_time"]
+
+		self.lag = obj.get("lag")
+		self.ping = obj["ping"]
+
+		self.version = obj["version"]
+		self.proto_min = obj["proto_min"]
+		self.proto_max = obj["proto_max"]
+
+		self.game_id = obj["gameid"]
+		self.url = obj.get("url")
+		self.name = obj["name"]
+		self.description = obj["description"]
+
+		if initial:
+			self.popularity = num_clients
+		else:
+			pop_factor = app.config["POPULARITY_FACTOR"]
+			self.popularity = self.popularity * pop_factor + \
+				num_clients * (1 - pop_factor)
+
+		self.geo_continent = obj.get("geo_continent")
+
+		self.creative = obj.get("creative", False)
+		self.damage_enabled = obj.get("damage", False)
+		self.pvp_enabled = obj.get("pvp", False)
+		self.password_required = obj.get("password", False)
+
+		self.last_update = now
+
+	def as_json(self):
+		obj = {
+			"address": self.address,
+			"can_see_far_names": self.can_see_far_names,
+			"clients_max": self.clients_max,
+			"clients_top": self.clients_top,
+			"creative": self.creative,
+			"damage": self.damage_enabled,
+			"dedicated": self.is_dedicated,
+			"description": self.description,
+			"game_time": self.game_time,
+			"gameid": self.game_id,
+			"name": self.name,
+			"password": self.password_required,
+			"ping": self.ping,
+			"pop_v": self.popularity,
+			"port": self.port,
+			"proto_max": self.proto_max,
+			"proto_min": self.proto_min,
+			"pvp": self.pvp_enabled,
+			"rollback": self.rollback_enabled,
+			"uptime": (datetime.utcnow() - self.start_time).total_seconds(),
+			"version": self.version,
+		}
+
+		# Optional fields
+		if self.clients is not None:
+			obj["clients_list"] = self.clients.split("\n") if self.clients else []
```

Comment:
If it's not none then split should already return an empty array
same with `mods` below

**Also**, since the JS presentation code relies on clients_list being present this should be set unconditionally.

Reply:
The empty string check is actually necessary because split() behaves differently if you pass it a delimiter:
```
>>> ''.split('\n')
['']
>>> ''.split()
[]
```